### ISSUE TITLE
Bipolar values should be displayed with explicit sign character

### ISSUE THREAD

**Issue by KlausStrohhaecker** [nonlinear-labs-dev/C15#3199]
At the moment, positive values (>0) of bipolar parameters etc do not show a "+" sign in the displays, whereas a "-" is displayed when they are negative. This is inconistent and causes a "jumpy" display.

![grafik](https://user-images.githubusercontent.com/42267360/171882761-9b2a0bb1-8e7e-467b-a0cd-39eb1152aa1e.png)
![grafik](https://user-images.githubusercontent.com/42267360/171882829-b9091e82-4e65-4593-b5e1-f7a47ec5d3d4.png)

Further, displaying an unconditional sign would help to easily tell bipolar values from unpolar ones (which don't need a sign)

**Comment by MatthiasSeeber** [comment]
> This is inconistent ...

I don't really see how. It is quite common to omit the technically redundant "+" and I am not familiar with any Synth >(or other device) displaying values in the proposed manner. This could be argued both ways, I suppose.

> ... and causes a "jumpy" display.

Fair point. Maybe, instead of a "+", a whitespace " " might also be possible to tackle this?

**Comment by KlausStrohhaecker** [comment]
It's a matter of style and intuitiveness, as well as personal preference of course.

I'd strongly prefer a _fixed_ length and explicit sign format for values from "`-100.0 %`"  to "`+100.0 %`". So we'd have `"  +3.2 %`" etc (no leading zeros on the integer part, rather use leading spaces). This will give very stable and benign display 
ballistics.

Whereas, making the whole number, and especially the decimal point jump around like crazy with varying numbers is definitely not the cleanest possible UI design (the free space on the display needs to be reserved anyway as we know the value range), and not something you'd ever do on a typical audio or measurement device, an industrial monitor/control device etc. Not moving targets (here: points of focus), please.

**Comment by MatthiasSeeber** [comment]
I don't disagree in general, but I still see flaws with the proposal.

- I would prefer focussing on negative Values in general (not distinguishing between uni- and bipolar Parameters, as described above).
- The proposal makes sense for linear Parameter Scalings, like the Bender as shown above. It would be rather overwhelming considering exponential Scalings like Envelope Times, which span from `0.000 ms`, `0.011 ms`, ... `16000 ms`, or Comb Decay (`0.000 ms`, ..., `100000 ms`). A fixed value Range for those would mean up to `+xxxxxx.xxx ms`.
- There is another prefix for returning HW Sources that needs to be considered: `!`. The images above don't show that, but when using the Encoder or Inc/Dec to change values, it will appear. This would be another character that has to be considered.
- The current rendering of C15 Oleds is coded manually (and may be comparable to drawing on a canvas). An implementation may be rather difficult.

**Comment by KlausStrohhaecker** [comment]
Obviously, for "log-space" variables like time, the floating (sic) point makes sense.
In general, I was about meaningful and optimized display for maximum readability.

For each numerical data type we want to display there is a known value range and other properties of the date (like the log-scaledness).

At the moment, the rendering functions into text strings don't get sufficient information passed to for discerning, for example, ranges for "%"-values wrt to signedness, sign explictness, sign position, leading spaces or zeroes, field widths etc.

I could fix that experimentally for LinearBipolar100PercentScaleConverter, though.

Another, and actually a bigger complication comes from the fact the font(s) for potentially "overprinted" numeric characters ('`0`'...'`9`', '`+`', '`-`' over '` `') is not fixed width as it should be (for consistent number display). A design failure ;-) And no easy fix, sadly.

**Comment by MatthiasSeeber** [comment]
Well, a general solution should aim for working with all Parameter scalings, not only linear ones. It would be strange i.m.o. to see well-aligned, highly readable values for percents, but not for milliseconds etc. Maybe, there are enough pixels for that.

Indeed, we would need to provide more descriptors for every Parameter on how to display it's scaled value (number of leading digits, trailing digits, can a value be negative).

We may be able to provide some special fixed-width unicode characters (or an extra font) for numeric display strings, solving the font issue. Still, tackling this by string manipulation seems not trivial and may require a well prepared concept/how-to.

**Comment by MatthiasSeeber** [comment]
> Additional fixed width characters (" ", "+" and "-", decimal point maybe as well) in the relevant fonts would be my preferred solution. We'd simply need to replace the original characters after conversion into string... in the ASCII domain (as that's the output), not unicode.

- Symbols we need in fixed width: space " ", plus "+", minus "-", dot "." and digits [0-9]
- In the ASCII domain, there is no room for _additional_ characters, and we don't want to overwrite anything. These original symbols are still needed for displaying parameter labels, info texts, etc.
- Concludingly, additional characters require Unicode (or another extended format beyond ASCII). By the way, Unicode replacements are already happening throughout the UI (for example - the special characters for Macro Labels A...F).

I don't mind experiments with the code, but I think there is benefit in having a discussion/meeting about the strategy before really implementing anything. The descriptors should be part of the `param-db` project (currently realized via spreadsheet, yaml experiments for a better solution are on their way).

To me, an important question still is, where this concept is applicable. Obviously, the main goal is the display in a typical Parameter Screen - and there seems to be some usable space:

![](https://www.nonlinear-labs.de/support/help/docs/HTML-Manual/images/hardware-ui/boled/parameter/boled-hw-adjust-src-pos.png)

However, positions are also displayed in other Screen modes where there is considerably less usable space:

![](https://www.nonlinear-labs.de/support/help/docs/HTML-Manual/images/hardware-ui/boled/parameter/boled-param-mc-amt.png)

![](https://www.nonlinear-labs.de/support/help/docs/HTML-Manual/images/hardware-ui/boled/parameter/boled-param-mod-range.png)

So it seems unlikely to find a real generic solution, at least for the Boled.

**Comment by KlausStrohhaecker** [comment]
Looking at the font data I'm puzzled... we are using bitmap fonts but they are **not** defined as bitmap fonts in the data (that's what the FontForge ttf Editor says and shows). Why, oh why? This doesn't make any sense at all and the fonts are almost impossible to edit.

**Comment by KlausStrohhaecker** [comment]
I don't see any problems here. We (as humans) do exactly know -- that is, we can exactly define -- how we would like to have the different outputs rendered with all the different constraints... and so can the software.

There is enough space in the 8-bit ASCII (standard C unsigned character) domain left for additional characters, at least for the fonts I've inspected so far. I'd simply use the orignal code +0x80, so for example 0x2B ("+") becomes 0xAB etc. None of the resulting codes are occupied as of yet.

**Comment by MatthiasSeeber** [comment]
I just said that it is unlikely to find a general solution. So we likely will display numeric values differently according to context. Sometimes we have enough space and can use it for well-aligned strings. Sometimes we haven't and need to fall back on the existing solution.

So, for the User and the Developer, we will introduce new inconsistencies and special rules depending on context. Yes, we humes do exactly know, but human minds also have their limitations regarding complex systems.

Regarding the char codes: why introduce a new special character space when we already have one?

https://www.nonlinear-labs.de/support/help/docs/HTML-Manual/developer/man-developer/index.html

The image above does only show one symbol variant of the concerned characters. So, they were overwritten? We still need the original variable-width symbols for label/info texts.

Can we please coordinate? So many things, designs and concepts are in question right now and I don't think a quick one-person-fix will do the job..

**Comment by KlausStrohhaecker** [comment]
I think we are overcomplicating things.
I've worked for some years in UI design (for industrial control systems) and each and every numeric output field on the screen had an explicit format specifier (in the data defining the screen layout and static and dynamic graphical elements, not the code) plus rules for corner cases where one single simple format specifier would not do. The rules actually were written in a PASCAL-like pseudocode which was "compiled" at program startup. This was 1990 or so (on 80486 DOS machines running Turbo Pascal)

35 years later, with C++ and all, I can't image it being any harder than it was back then.

> Can we please coordinate? So many things, designs and concepts are in question right now and I don't think a quick one-person-fix will do the job..

Agreed. I would think we might not tackle the issue in the C15 but we should keep this topic in mind for the C16, not repeating the same flaws there, @hhoegelo

**Comment by MatthiasSeeber** [comment]
> I think we are overcomplicating things.

The ticket started with the requirement: "Bipolar values should be displayed with explicit sign character" - which is still the title btw. Then, it's about well-aligned strings and monospace, "design failures" seem to be everywhere. Fonts need to be tempered with, we now introduce another special character region. The core of the UI, with all these Screens with their special rules, needs enhancements. Yes, this got complicated quite quickly. And I find it unfortunate to have the concerns I raised being dismissed.

Of course I don't question your experience with UI design. However, designing UIs (from scratch) and building on an already existing design with its limitations, are two different things. In my seven years working here, there were lots of tasks with seemingly simple UI changes that turned out to be orders of magnitude more complicated.

If this is more focussed on the C16, then - as I already mentioned - we could exploit HTML (as the UI is intended to be rendered this way) - which will be a lot easier (presumably) and more flexible (almost certainly) than string manipulation. The additional benefit would be the possibility to try out new designs (with better readability in mind from the start) and maybe even finding a _general_ solution which works on every/any Screen mode.
